### PR TITLE
[transaction][acknowledge] Persist pending ack messages to cursor meta store.

### PR DESCRIPTION
**Pull request by MarvinCai**
Add message SubscriptionPendingAckMessages to store pending ack message positions.
Add update/getSubscriptionPendingAckMessages to managed ledger MetaStore for storing/getting pending ack message positions.
Persisst pending ack message positions when acknowledging message and recover pending ack message positions when initializing subscription.

### CODE REVIEW

**Comment by sijie** [comment]
@MarvinCai the overall logic looks good to me. Can you add tests for it?

**Comment by MarvinCai** [comment]
@sijie I finished the code change and added test cases, can you take another look, thanks!

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java [hunk 1]
```diff
@@ -2015,6 +2020,59 @@ public void operationFailed(MetaStoreException e) {
...
+
+    public void getPendingAckPositionMetaInfo(String subName,
+                                              MetaStoreCallback<SubscriptionPendingAckMessages> callback) {
+        if (isClosed()) {
+            callback.operationFailed(new MetaStoreException(
+                    new ManagedLedgerException.CursorAlreadyClosedException(name + " cursor already closed")));
+            return;
+        }
+
+        ledger.getStore().asyncGetSubscriptionPendingAckMessages(ledger.getName(), name, subName,
```
Comment (sijie) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java]: interesting... why do you need to create a new callback? Can't you just use the passed-in callback?

```
ledger.getStore().asyncGetSubscriptionPendingAckMessages(ledger.getName(), name, subName, callback);
```

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java [hunk 2]
```diff
@@ -2015,6 +2020,59 @@ public void operationFailed(MetaStoreException e) {
...
+            callback.operationFailed(new MetaStoreException(
+                    new ManagedLedgerException.CursorAlreadyClosedException(name + " cursor already closed")));
+            return;
+        }
+
+        SubscriptionPendingAckMessages.Builder subscriptionPendingAckMessageBuilder = SubscriptionPendingAckMessages.newBuilder()
+                    .setPendingCumulativeAckMessagePosition(pendingCumulativeAckMessagePositionBuilder)
+                    .addAllPendingAckMessages(pendingAckMessageEntryBuilderList);
+
+        log.debug("[{}] [{}] [{}] Persisting subscription's pending ack messages to meta-data store",
```
Comment (sijie) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java]: wrap this into `log.isDebugEnabled()`.

```
if (log.isDebugEnabled()) {
    ...
}
```

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java (hunk 1 above)
Comment (MarvinCai) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java]: updated.

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java (hunk 2 above)
Comment (MarvinCai) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java]: updated.

**Comment by MarvinCai** [comment]
Updated ManagedCursorImpl to add functionality to persist pending ack messages info to a separate ledger.
Ledger will be created when recover or start to persist pending ack messages info.
If ledger creation failed an internal state will be set to NoLedger.

When asyncUpdatePendingAckPosition API is invoked, it'll check the state, if it's in Open will try to persist  ack messages info to ledger, else it'll persist to ledger meta store and try to create new ledger.

Subscription will also first try to do asyncUpdatePendingAckPosition then update the in memory status it keep about ongoing txns pending ack messages.

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java [hunk 3]
```diff
@@ -426,6 +454,85 @@ public void operationFailed(ManagedLedgerException exception) {
...
+                    return;
+                } else {
+                    // Pending ack message info is stored in ledger.
+                    ledger.mbean.startCursorLedgerOpenOp();
+                    long ledgerId = subscriptionPendingAckMessages.getPendingAckMsgLedgerId();
+                    bookkeeper.asyncOpenLedger(ledgerId, digestType, config.getPassword(), (rc, lh, ctx) -> {
+                        if (isBkErrorNotRecoverable(rc)) {
+                            log.error("[{}] Cursor {} fail to open pending ack message ledger {} for recover",
+                                    ledger.getName(), name, ledgerId);
+                            startCreateNewPendingAckMsgLedger();
```
Comment (sijie) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java]: If we fail to open the ledger, why do we call `startCreateNewPendingAckMsgLedger` here?

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java [hunk 4]
```diff
@@ -426,6 +454,85 @@ public void operationFailed(ManagedLedgerException exception) {
...
+                            future.completeExceptionally(createManagedLedgerException(rc));
+                            return;
+                        } else if (rc != BKException.Code.OK) {
+                            log.error("[{}] Cursor {} fail to open pending ack message ledger {} for recover",
+                                    ledger.getName(), name, ledgerId);
+                            startCreateNewPendingAckMsgLedger();
```
Comment (sijie) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java]: same question here as above.

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java [hunk 5]
```diff
@@ -426,6 +454,85 @@ public void operationFailed(ManagedLedgerException exception) {
...
+                            future.completeExceptionally(createManagedLedgerException(rc));
+                            return;
+                        }
+
+                        // Read the last entry in the pending ack message ledger
+                        long lastEntryInLedger = lh.getLastAddConfirmed();
+
+                        if (lastEntryInLedger < 0) {
+                            log.warn("[{}] Cursor {} fail to read from pending ack message ledger {} for recover: " +
```
Comment (sijie) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java]: I don't think this should be a failure. it just means that it is an empty ledger, right?

**Comment by codelipenghui** [comment]
move to 2.7.0

**Comment by codelipenghui** [comment]
This task will be handled by https://github.com/apache/pulsar/pull/8295, So I will close this PR first.

**Pull request closed without merging**